### DISCUSSION TITLE
Azure batch : Tasks == Cores

### DISCUSSION THREAD

**Discussion by ossonts** [nextflow-io/nextflow#3305] [Q&A]
https://github.com/nextflow-io/nextflow/blob/904c9409d25c64236a48962c95625666e6c57aca/plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzBatchService.groovy#L516

Anyone know why this is enforced? Surely you would not want the tasks to be the same as the core count?

**Comment by pditommaso** [comment]
Because nextflow assumes 1 cpus == 1 slot to determine the slots requirement for each task 

https://github.com/nextflow-io/nextflow/blob/904c9409d25c64236a48962c95625666e6c57aca/plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzBatchService.groovy#L362-L372

**Reply by ossonts** [reply]
Thanks, I am hoping to alter this, as I don't want to use all cores, say 10 of 12 the cores used for tasks -- are then implications to changing this?

**Reply by pditommaso** [reply]
You can specify in the task (or in the config file) how many cpus you want to use using the [cpus](https://www.nextflow.io/docs/latest/process.html#cpus) directive. You don't to hack in the code.

**Reply by ossonts** [reply]
yes but if I have a machine with 120 cores how do I specify that I only want 100 of the cores to be running my tasks?

**Reply by pditommaso** [reply]
Nextflow is designed to request resources per task level. If you are planning to cap the amount of cpus per node is not possible. 

Instead, if your requirement is to cap the top amount of cpus used by a workflow execution you can achieve it using the configuration setting `executor.queueSize`. 

For example, having a task using 2 cpus, you can set `executor.queueSize = 50`.

**Comment by ossonts** [comment]
Sorry, I am not sure I follow. I have 10 machines with 120 cores (total of 1200 cores) -- Lets say all my tasks are 1 core each.  In batch I make my own pool and limit the slots per node to 100 (leaving 20 cores for other stuff).  So I can run 1000 jobs in parallel. 
Is it not possible to read the slots per node (100) X nodes (10) and tell nextflow to schedule appropriately? (are you around for a chat sometime?)

**Comment by ossonts** [comment]
So I ran a quick test.... 

I pre-made a pool with: (Azure portal)
2 x standard_d15_v2 -- these are 20 cores each 
on the pool I set the number of tasks per VM to 16 (leaving a few cores spare)

Then in the nextflow azure config, 
```
 vmType = 'Standard_D14'

```
Note this is a 16 core machine -- **NOT** the actual one I am using -- I set this as the cores on this VM == 16

Then I ran 3 jobs with 1000 tasks each. 

VM count ==1 and then part way through set the  VM count to 2. 

So with nextflow thinking I have 16 cores VMs and azure limiting to 16 per VM, I see the expected 32 tasks.

![image](https://user-images.githubusercontent.com/472243/198229447-c4d2979f-aaec-4acc-84e1-879a874a6d5b.png)

Somehow I feel that for a predefined pool (one created in the Azure Portal) -- I should just be able to set the batch credentials and then on each nextflow process set the number of cores that it should use (my case 1)  -- having to specify a VM type for a predefined pool seems a bit obsolete -- especially as there is no validation..... 

happy to take any thoughts....

**Comment by ossonts** [comment]
how about this:

If the pool is preconfigured (azure portal , with a predefined number of tasks per node)

Can nextflow not just use this task count ?
